### PR TITLE
Feature: status reporting for ALB releases

### PR DESCRIPTION
**Pull request by izaaklauer**
This implements status reports for the aws alb releaser plugin. It waits briefly for the release's target group to finish initializing, then reports on the health of the targets.

A few caveats:
- For applications that are slow to start (and most ec2 deployments), the initial status check will return unhealthy or partial. 
- I'm not looking at target group health. Currently not reporting on the state of the ALB, security group, DNS records, or any other resources at play during a release.
- This assumes we only have one target group associated with each release. When we implement more complex deployment strategies that require having multiple target groups behind the same ALB, this will need to be re-thought.
- Lambda function target groups have health checks disabled. Rather than present state UNKNOWN for all lambda releases, I'm mapping disabled to healthy. If that disabled state shows up in any other scenarios, we may need to revisit this.
- I'm not nil-checking before dereferencing pointers returned by the aws api, in keeping with our existing patterns.
 
Tested with deployments from ec2 and lambda.

### CODE REVIEW

File: builtin/aws/alb/releaser.go [hunk 1]
```diff
@@ -304,9 +315,145 @@ func (r *Releaser) Release(
 	return &Release{
 		Url:             "http://" + hostname,
 		LoadBalancerArn: *lb.LoadBalancerArn,
+		TargetGroupArn:  target.Arn,
+		Region:          target.Region,
```
Comment (izaaklauer) [inline builtin/aws/alb/releaser.go:319]: I needed the region to build the right client. [Other plugins](https://github.com/hashicorp/waypoint/blob/main/builtin/aws/ec2/platform.go#L95) have this on their config. Not sure which pattern we prefer.

File: builtin/aws/alb/releaser.go [hunk 2]
```diff
@@ -304,9 +315,145 @@ func (r *Releaser) Release(
...
 
+func (r *Releaser) Status(
+	ctx context.Context,
+	log hclog.Logger,
+	release *Release,
+	ui terminal.UI,
+) (*sdk.StatusReport, error) {
+
+	if release.Region == "" {
+		log.Debug("Region is not available for this release. Unable to determine status.")
```
Comment (izaaklauer) [inline builtin/aws/alb/releaser.go:331]: I have a feeling that we might hit this case in the upgrade path, where someone upgrades a server then tries to promote a deployment that didn't encode the region?

**Review by krantzinator — APPROVED** [review]
I haven't fully tested this, but the basics look good.

File: builtin/aws/alb/releaser.go (hunk 2 above)
Comment (krantzinator) [reply builtin/aws/alb/releaser.go:331]: This would be good to test, sounds like a correct assumption.

File: builtin/aws/alb/releaser.go [hunk 3]
```diff
@@ -304,9 +315,145 @@ func (r *Releaser) Release(
...
+		})
+	}
+
+	// If AWS registers targets slowly or incrementally, we may report an artificially low number of total targets.
+	totalTargets := len(targetHealthDescriptions)
+
+	if healthyCount == totalTargets {
+		report.Health = sdk.StatusReport_READY
+		msg := fmt.Sprintf("All %d targets are healthy.", totalTargets)
+		report.HealthMessage = msg
```
Comment (briancain) [inline builtin/aws/alb/releaser.go:441]:
```suggestion
		report.HealthMessage =  fmt.Sprintf("All %d targets are ready.", totalTargets)
		st.Step(terminal.StatusOk, report.HealthMessage)
```

File: builtin/aws/alb/releaser.go [hunk 4]
```diff
@@ -304,9 +315,145 @@ func (r *Releaser) Release(
...
+	} else if healthyCount > 0 {
+		report.Health = sdk.StatusReport_PARTIAL
+		msg := fmt.Sprintf("%d/%d targets are healthy.", healthyCount, totalTargets)
+		report.HealthMessage = msg
+		st.Step(terminal.StatusWarn, msg)
+	} else {
+		report.Health = sdk.StatusReport_DOWN
+		msg := fmt.Sprintf("All targets are unhealthy, however your application might be available or still\nstarting up.")
+		report.HealthMessage = msg
+		st.Step(terminal.StatusWarn, msg)
```
Comment (briancain) [inline builtin/aws/alb/releaser.go:451]:
```suggestion
		report.HealthMessage =  "All targets are unhealthy, however your application might be available or still starting up."
		st.Step(terminal.StatusWarn, report.HealthMessage)
```

File: builtin/aws/alb/releaser.go [hunk 5]
```diff
@@ -304,9 +315,145 @@ func (r *Releaser) Release(
...
+	report.Resources = []*sdk.StatusReport_Resource{}
+
+	healthyCount := 0
+	for _, tgHealth := range targetHealthDescriptions {
+
+		targetId := *tgHealth.Target.Id
+
+		var health sdk.StatusReport_Health
+
+		switch *tgHealth.TargetHealth.State {
```
[the remark this answers is not in the document]
Comment (briancain) [reply builtin/aws/alb/releaser.go:408]: Maybe leave a note above the switch?

**Review by briancain — CHANGES_REQUESTED** [review]

File: builtin/aws/alb/releaser.go [hunk 6]
```diff
@@ -304,9 +315,145 @@ func (r *Releaser) Release(
...
+	startTime := time.Now().Unix()
+	for {
+		if targetHealthDescriptions != nil {
+			sleepDuration := time.Second * time.Duration(targetGroupInitializationPollingIntervalSeconds)
+			log.Debug("Sleeping %0.f seconds to give the following target group time to initialize:\n%s", sleepDuration.Seconds(), release.TargetGroupArn)
+			time.Sleep(sleepDuration)
+		}
+
+		if startTime+int64(targetGroupInitializationTimeoutSeconds) <= time.Now().Unix() {
+			return nil, fmt.Errorf("timed out after %d seconds waiting for the following target group to initialize:\n%s", time.Now().Unix()-startTime, release.TargetGroupArn)
```
[the remark this answers is not in the document]
Comment (briancain) [reply builtin/aws/alb/releaser.go:363]: Or even better just mark the status report as Unknown and break out of the loop completely and then the health message could be that the initialization timed out after `x` seconds.

File: builtin/aws/alb/releaser.go [hunk 7]
```diff
@@ -12,14 +11,21 @@ import (
 	"github.com/hashicorp/go-hclog"
 	"github.com/hashicorp/waypoint-plugin-sdk/component"
 	"github.com/hashicorp/waypoint-plugin-sdk/docs"
+	sdk "github.com/hashicorp/waypoint-plugin-sdk/proto/gen"
 	"github.com/hashicorp/waypoint-plugin-sdk/terminal"
 	"github.com/hashicorp/waypoint/builtin/aws/utils"
+	"time"
```
[the remark this answers is not in the document]
Comment (krantzinator) [reply builtin/aws/alb/releaser.go:17]: I have a plugin that auto-arranges these for me. I'm actually not sure if it's the base Go extension or something I added separately.

[the remark this answers is not in the document]
Comment (briancain) [reply builtin/aws/alb/releaser.go:17]: Same... mine auto-formats!

**Review by briancain — APPROVED** [review]
Code looks good to me! 🎉

**Pull request merged**